numpy/_core/src/multiarray/item_selection.c

```diff
@@ -2868,6 +2905,35 @@ PyArray_Nonzero(PyArrayObject *self)
     if (ret == NULL) {
         return NULL;
     }
+    /* nothing to do */
+    if (nonzero_count == 0) {
+        goto finish;
+    }
+
+    PyArrayObject* original_array = self;
+    if (PyArray_BASE(self) != NULL) {
+        original_array = (PyArrayObject*) PyArray_BASE(self);
+    }
+    int bytes_not_swapped = PyArray_ISNOTSWAPPED(self) && PyArray_ISNOTSWAPPED(original_array);
+
+    // do not add ndim=1, dtype->kind == 'b', since we have a separate fast path for it
+    int optimized_count = PyArray_TRIVIALLY_ITERABLE(self) && ! (ndim == 1 && dtype->kind=='b') && bytes_not_swapped;
+    if (optimized_count ) {
+        npy_intp * multi_index = (npy_intp *)PyArray_DATA(ret);
+        char * data = PyArray_BYTES(self);
+        const npy_intp M_dim = PyArray_NDIM(self);
+        const npy_intp* M_shape = PyArray_SHAPE(self);
+        const npy_intp* M_strides = PyArray_STRIDES(self);
+        int M_type_num = dtype->type_num;
+
+        bool executed = nonzero_idxs_dispatcher((void*)data, multi_index, M_dim,
+                                        M_shape, M_strides, M_type_num, nonzero_count);
```

Comment:
This branch is missing the thresholded GIL release (internals of it I guess).

Reply:
Hmm. If we release the GIL and the underlying array is changed so there are more zero-elements, then we can get a buffer overflow. I will try to pass the allocated array size and add a guard in `nonzero_idxs_dispatcher`

Reply:
Releasing the GIL (or running the cpython free-threading build) could result in a buffer overflow if another thread changes the number of non-zero entries in the array. I added a guard for that, but we should rerun the benchmarks (I still expect a good speedup though).

Should a test be added for this? (probably here: https://github.com/numpy/numpy/blob/main/numpy/_core/tests/test_multithreading.py)

Reply:
Interesting, a case where we would actually want to lock down a whole array (but I think this is super hard with buffer exchange, so in practice...).

Would be very cool if you add a test.  Larger use of free-threading is in the future after-all, and maybe not a too distant one.

Reply:
Tests have been added in https://github.com/numpy/numpy/pull/28361